File: 004 - Largest palindrome product/ProjectEuler+ Challenge #4 Solution.c
#include <stdio.h>

int main() {
    int T, N, largest, limit, start, step, temp, product, reverse;
    scanf("%d", &T);
    for (int _ = 0; _ < T; _++) {
        scanf("%d", &N);
        largest = 0;
        limit = N - 1;
        for (int num_1 = 999; num_1 > 100; num_1--) {
            if (num_1 % 11 != 0) {
                temp = limit / num_1 / 11 * 11;
                start = 990 < temp ? 990 : temp;
                step = -11;
            } else {
                temp = limit / num_1;
                start = 999 < temp ? 999 : temp;
                step = -1;
            }
            for (int num_2 = start; num_2 > num_1; num_2 += step) {
                product = num_1 * num_2;
                if (product <= largest) {
                    break;
                }
                reverse = 0;
                temp = product;
                while (temp != 0) {
                    reverse *= 10;
                    reverse += temp % 10;
                    temp /= 10;
                }
                if (reverse == product) {
                    largest = product;
                    break;
                }
            }
        }
        printf("%d\n", largest);
    }
    return 0;
}

File: 004 - Largest Palindrome Product/ProjectEuler+ Challenge #4 Solution.py
T = int(input())
for _ in range(T):
    N = int(input())
    largest = 0
    limit = N - 1
    for number_1 in range(999, 100, -1):
        if number_1 % 11:
            start = min(990, limit // number_1 // 11 * 11)
            step = -11
        else:
            start = min(999, limit // number_1)
            step = -1
        for number_2 in range(start, number_1, step):
            product = number_1 * number_2
            if product <= largest:
                break
            product_str = str(product)
            if product_str == product_str[::-1]:
                largest = product
                break
    print(largest)
